src/snowflake/connector/cache.py

```diff
@@ -260,3 +319,218 @@ def _expiration(self, k: K) -> None:
+                if loaded:
+                    t, v = self._cache[k]
+                    if not is_expired(t):
+                        expire_item = False
+                if expire_item:
+                    self._expiration(k)
+                    self._delitem(k)
+                    raise KeyError
+            self._hit(k)
+            return v
+        finally:
+            if currently_holding:
+                self._lock.release()
+
+    def _setitem(self, k: K, v: V) -> None:
+        super()._setitem(k, v)
+        self._save_if_should()
+
+    def _load(self) -> bool:
+        """Load cache from disk if possible, returns whether it was able to load."""
+        try:
+            with open(self.file_path, "rb") as r_file:
+                other = pickle.load(r_file)
+            _ = self._update(
+                other,
+                update_newer_only=True,
+            )
+            self.last_loaded = now()
+            return True
+        except OSError:
+            return False
+
+    def _save(self) -> bool:
+        """Save cache to disk if possible, returns whether it was able to save."""
+        self._clear_expired_entries()
+        try:
+            with self._file_lock:
+                self._load_if_should()
+                _dir, fname = os.path.split(self.file_path)
+                # TODO add exception handling for folder not being writeable
+                tmp_file, tmp_file_path = tempfile.mkstemp(
+                    prefix=fname,
+                    dir=_dir,
+                )
+                with open(tmp_file, "wb") as w_file:
+                    pickle.dump(self, w_file)
+                # We write to a tmp file and then move it to have atomic write
+                os.replace(tmp_file_path, self.file_path)
+                self.last_loaded = datetime.datetime.fromtimestamp(
+                    os.path.getmtime(self.file_path),
+                )
+                return True
+        except Timeout:
+            logger.debug(
+                f"acquiring {self._file_lock_path} timed out, skipping saving..."
+            )
+        return False
+
+    def _save_if_should(self) -> bool:
+        """Saves file to disk if necessary and returns whether it saved.
+
+        Uses self._should_save to decide whether to save.
+        """
+        if self._should_save():
+            return self._save()
+        return False
+
+    def _load_if_should(self) -> bool:
+        """Load file to disk if necessary and returns whether it loaded.
+
+        Uses self._should_load to decide whether to load.
+        """
+        if self._should_load():
+            return self._load()
+        return False
+
+    def _should_save(self) -> bool:
+        """Decide whether we should save.
+
+        This is a simple random number generator to randomize writes across processes
+        that are possibly saving the same values in this cache.
+        """
+        return random.randint(0, 9) == 1
```

Comment:
Do we want to make this `9` a static variable? I guess we don't need to change this a lot but in case it is needed it's easier to adjust.

Reply:
Sure, this is probably a good idea